<p>The genetic code of all organisms uses a 3 base (codon), 4 letter encoding (A, G, C or T/U) to represent the 20* amino acids used in proteins. This yields 4<sup>3</sup> = 4*4*4 = 64 different possible three base codons. Of these, one is used as an initiator called "start codon", three are used to signal the end of a protein and are called "stop codons" (*). The residual 60 codons + the start/methionine codon encode the 20 proteinogenic amino acids. Some amino acids are encoded by up to 6 different codons, whereas other amino acids are only encoded by a single codon. This is known as the <em><a href="https://en.wikipedia.org/wiki/Codon_degeneracy" target="_blank">degenerate code</a></em> and is often visualized by a <a href="https://upload.wikimedia.org/wikipedia/commons/7/70/Aminoacids_table.svg" target="_blank">codon wheel</a>. Every organism has a different set of <a href="https://en.wikipedia.org/wiki/Codon_usage_bias" target="_blank">preferred codons</a> which helps to optimize and balance protein production.</p>

<p>In this bite you are provided with a list of all coding sequences of the bacterium <em><a href="https://en.wikipedia.org/wiki/Staphylococcus_aureus">Staphyloccocus aureus</a></em>.</p>

<p>Calculate the <a href="https://www.kazusa.or.jp/codon/cgi-bin/showcodon.cgi?species=426430&aa=11&style=N" target="_blank">average codon usage table</a> for all sequences using the supplied translation table. Please note that the coding sequences are supplied a an RNA sequence, whereas the codon usage table is provided as a DNA sequence. To convert a DNA sequence to an RNA sequence, replace all <em>T</em>s to <em>U</em>s. Disregard sequences that are not valid coding sequences.</p>

<hr>

<p>There you go, our first <i>Bioinformatics</i> Bite. Keep calm and code in Python!</p>